### ISSUE TITLE
"The namespace or module '...' is not defined" despite module existing

### ISSUE THREAD

**Issue by CanePlayz** [ionide/ionide-vscode-fsharp#1840]
### Describe the bug

There is syntax highlighting with the error message of "The namespace or module '...' is not defined" despite the module existing in that exact version in my local files. When I build the program or run the tests (the file with the error is my test.fs file), everything works fine, but I get this annoying syntax highlighting.

### Steps to reproduce

1. Have a file that imports packages.

### Expected behaviour

No syntax highlighting

### Screenshots

Module highlighting:
![image](https://user-images.githubusercontent.com/81419612/224155799-a27e9857-5115-41de-bc4f-71c368612d68.png)

Resulting highlighting:
![image](https://user-images.githubusercontent.com/81419612/224156180-4b484cf8-4827-42b1-98b1-1cf3b5807f72.png)

## Machine info
 - OS: Windows 11
 - .NET SDK version: Appears on 6.0.309 and 7.0.201
 - Ionide version: 7.5.1

### Additional context
If you want the exact files, I can share them with you. Just let me know.

**Comment by MangelMaxime** [comment]
Hello @CanePlayz,

Does the error goes away if you quit and restart VSCode?

As a workaround can you try the following:

**Workaround 1**

1. Save/Commit the work you don't to lose
2. Clean your temporary files `git clean -xdf`
3. Run `dotnet build` to try forcing Ionide to update

**Workaround 2**

1. Save/Commit the work you don't to lose
2. Clean your temporary files `git clean -xdf`
3. Quit and restart Ionide

Sometimes I need to do stuff like that on some of my projects.

**Comment by CanePlayz** [comment]
> Does the error goes away if you quit and restart VSCode?

Unfortunately, no. Sometimes, it does find the modules. And something that is different in those scenarios is that there is a file called something like NET Testing SDK...However, that might just be a side effect of the project finding the file, because that file only gets auto-generated when the module is found, and not the other way around (the file helping Ionide).

Thanks for the workarounds, however, they don't apply in this case since this is not a Git project. It's basically many very tiny projects, just exercises from university, only consisting of a .fsproj. and two to three .fs files.

**Comment by MangelMaxime** [comment]
The clean command is because in general you add the `bin` and `obj` folders to the gitignore file.

If you don't use Git, you can always remove the `bin` and `obj` folders manually or using the CLI.

**Comment by CanePlayz** [comment]
> The clean command is because in general you add the `bin` and `obj` folders to the gitignore file.
> 
> If you don't use Git, you can always remove the `bin` and `obj` folders manually or using the CLI.

Yeah, thanks, I do that anyways on a regular basis with a little script.

**Comment by CanePlayz** [comment]
Do you want one of those projects to see if you can replicate the issue?

**Comment by MrLuje** [comment]
I had the same issue with the default SAFE template, enabling `Adaptive Lsp Server` fixed it for me (https://github.com/ionide/ionide-vscode-fsharp/issues/1852#issuecomment-1469005519)

**Comment by CanePlayz** [comment]
Issue still persists with Adaptive Lsp Server enabled, unfortunately. It's also weird that on some of the exercise projects, it finds the modules, despite them not being any different from the other ones as far as I can see.

**Comment by CanePlayz** [comment]
After running "dotnet test" once, it finds the modules in the test files.

**Comment by CanePlayz** [comment]
> Can you submit a repository that reproduces this issue?

You can find the project here: https://github.com/CanePlayz/Ionide-Test

**Comment by TheAngryByrd** [comment]
After a dotnet restore I cannot get this to reproduce.

https://user-images.githubusercontent.com/1490044/227225283-f949fc26-fa1e-44b6-a9fc-3a381781ea0b.mp4

---

Maybe there's something weird about your install of dotnet. What's `dotnet --info` ?

**Comment by CanePlayz** [comment]
> After a `dotnet restore` I cannot get this to reproduce.

Me neither anymore. Maybe something fixed it. It always finds the modules now after a `dotnet restore`. But why does it need that `dotnet restore` if the packages are already installed?

> What's `dotnet --info`?

```
.NET SDK:
 Version:   7.0.202
 Commit:    6c74320bc3

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.25314
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.202\

Host:
  Version:      7.0.4
  Architecture: x64
  Commit:       0a396acafe

.NET SDKs installed:
  6.0.407 [C:\Program Files\dotnet\sdk]
  7.0.202 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by baronfel** [comment]
Restore doesn't just download nuget packages, more importantly it creates several metadata files (like project.assets.json) in the `Intermediate Output Path` of each project that are required in order to provide intellisense for the files in that project. The fact that performing a restore outside of the editor cleared things up for you makes me think that our automatic restore-triggering in FSAC isn't working consistently.

**Comment by davedawkins** [comment]
Thank you for this thread. I've been losing my mind. "dotnet restore".

**Comment by CanePlayz** [comment]
@MangelMaxime I cannot reproduce this anymore. Should we close this issue?

**Comment by davedawkins** [comment]
I still get this issue. It's always after I add a new `.fs` file to my `.fsproj`. The reported error is usually to do with an included project like this:

```
    <ProjectReference Include="../../../sutil/src/Sutil/Sutil.fsproj"/>
```

I would discount installation of .NET etc because it still happens even after installing a brand new laptop (the thinking being that having followed standard installation procedures on a new machine, there's been little to no chance for the .net install to get funky).

I'd be happy to try and make a repro project if that would help?

**Comment by CanePlayz** [comment]
Yeah, I also just remembered that this was not just about modules not being found in general, but also about not finding them before doing a manual `dotnet restore`, which is still necessary but shouldn't be.

**Comment by davedawkins** [comment]
For what it's worth, I'm still needing to do "dotnet restore" (sometimes intermixed with "Clear Project Cache", though may not be necessary) and "Reload Window", most frequently after making a change to an .fsproj file. 

```
MacOS: Sequoia 15.3.2
dotnet: 8.0.406
Ionide: 7.26.1
```

It's manageable and part of my life. I could try and help if someone can point me at general area of the code-base...

**Comment by TheAngryByrd** [comment]
We do our [design time](https://github.com/dotnet/project-system/blob/main/docs/design-time-builds.md) project loading in [FSAC here](https://github.com/ionide/FsAutoComplete/blob/a8a3a9ac40f9a04d0d4985152b67af8b87e791f6/src/FsAutoComplete/LspServers/AdaptiveServerState.fs#L907). I don't know if there's a way to detect if you need to restore or not. 

The one problem I'll immediately point out that occurred in the past is collisions with other `dotnet restore/build/etc` tasks. We currently use the same obj/ folder as `dotnet restore` and we _can't_ know about other processes. So if some outside build script causes us to trigger a design time build and we call `dotnet restore` at the same time as another dotnet process doing things, one of them might fail for non-obvious reasons, especially on Windows with its file locking.

Because of that we'd probably want to restore and do design time builds based on some other `tmp` directory. You could do this by setting `BaseIntermediateOutput` property during restore and design time builds. I think anything we rely on would be fine with this _but_ not every msbuild task/target utilizes this property correctly and still might not restore properly for some people.

**Comment by CanePlayz** [comment]
@TheAngryByrd Thanks for the info! So are there any plans to fix this long term or should I close this issue?

**Comment by davedawkins** [comment]
My encounter with this error was fixed by renaming a module from Application.UI.Foo to just AppUI.Foo.

Compiler was ok just not Ionide. 

 I can try and repro on a different issue if you wanted to close this one.

**Comment by TheAngryByrd** [comment]
@CanePlayz 
> [@TheAngryByrd](https://github.com/TheAngryByrd) Thanks for the info! So are there any plans to fix this long term or should I close this issue?

Thanks! No plans but any contributions would be accepted.

@davedawkins 
> My encounter with this error was fixed by renaming a module from Application.UI.Foo to just AppUI.Foo.
> 
> Compiler was ok just not Ionide.
> 
> I can try and repro on a different issue if you wanted to close this one.

A repro with the structure that fails would be a tremendous start given how complex msbuild is to debut with out.

**Comment by CanePlayz** [comment]
@davedawkins That would be much appreciated. Thanks!
